### ISSUE TITLE
Feature Request: Support weighted products

### ISSUE THREAD

**Issue by gentoo-root** [grocy/grocy#2387]
Context: There are products sold in packs, each pack has an arbitrary weight (the barcode includes a mask that denotes the weight, e.g. 2788507**00307**3 is 307 g, see https://github.com/grocy/grocy/issues/1013). One pack is typically consumed at a time, while Calories are calculated per 100 g.

It would be great to be able to:

1. Track packs and grams at the same time. Track weight of each pack.
2. When adding a product, extract the grams from the barcode and add one pack with the corresponding weight.
3. When consuming a product, consume one pack.
3.1. Choose the pack by weight.
3.2. Extract the weight from the barcode and choose the corresponding pack.
4. Use grams for energy value and prices, use packs for quick consume.

**Comment by berrnd** [comment]
Generally Grocy is all about having conversions between different units, either global or product specific ones. So what have you tried so far as a concrete practical example and what does not work exactly by that ([persistent demo instances](https://www.reddit.com/r/grocy/comments/jwnk9b) are there share complex scenarios)?

As mentioned on the Feature Request template, please keep it to one topic per request - one Request for everything is not trackable.

Stock entry specific unit conversions are not likely to be added, since practically no one wants to enter this anyway. You can identify each individual stock entry (regardless of its unit(s)) by a barcode when utilizing stock entry labels.

&nbsp;
> Use grams for energy value and prices, use packs for quick consume.

Duplicate of #2252 and also have a look at allt hose product options:

![grafik](https://github.com/grocy/grocy/assets/2161815/983dbc84-d6b6-400c-9359-13a1d2488079)

**Comment by dillwishlist** [comment]
This sounds like an ideal way to simplify pack handling.  I would very much like to see this!

**Comment by berrnd** [comment]
>  I would very much like to see this!

Well, defining what exactly is not already possible by what I just mentioned (utilizing Userfields, taking all the existing product options into account - please have a look [at above](https://github.com/grocy/grocy/issues/2387#issuecomment-1818841177)) would be cool, I don't get it.

Please simply provide a reproducible practical example (=> [persistent demo instance](https://www.reddit.com/r/grocy/comments/jwnk9b)) to point that out, it's so much easier to look at a concrete example instead of getting lost in theory.

**Comment by berrnd** [comment]
> Stock entry specific unit conversions are not likely to be added

Just referencing another related thread on that: #2171

**Comment by berrnd** [comment]
Thanks for the practical example.

&nbsp;

> Weight decoding from barcodes is already tracked at https://github.com/grocy/grocy/issues/1013. Units for energy value are tracked at https://github.com/grocy/grocy/issues/2252, as you pointed out. The list above represents the use cases that would be possible if this was implemented.

So for that simply duplicates, no need to track that another time here.

&nbsp;
> It's essentially one topic

No, it's at least about the two duplicates + more. How this produces untrackable clutter is pretty much already visible right here.

&nbsp;
> and I can't set price per gram on the purchase page (probably deserves a separate issue?):

From a practical stand point: You bought 6 Packs and you know for sure what how much they were together. So is using "Total price" and entering simply that price not the easiest option of all? Based on unit conversions, Grocy can calculate/show the price in any other unit without problems.

As said, stock entry specific  unit conversions won't be added, since that would require to rewrite (once gain) the whole unit conversion system, which is already a way too complex part for what Grocy was originally intended for. The single price entered on [/purchase](https://demo-prerelease.grocy.info/purchase) is always related to the selected unit, everything else makes no sense / leads at least to even more confusion.

The just mentioned other example #2171 for that is practically about the same - of course a little difference for the average price and the like on varying-weighted products like fruits will happen, but this will be practically pretty irrelevant over time while still being pretty much correct.

&nbsp;
> I can't buy a pack of X grams, because unit conversions are global, and there is no way to specify the weight of an individual pack (this is the topic of this issue)

To clarify that just once more: Product specific unit conversions will not happen in Grocy due to what I just mentioned.

For this practical example: When you buy Borsch in Packs and even consume it in Packs, as you've mentioned above, what's even the practical point to track Grams at all? Similar example: A roll of toilett paper roughly weighs 95 Grams, but since consumption only happens per roll, tracking Grams makes not that much sense there either, at least it adds nothing additionally practical useful.

&nbsp;
> That requires hassle with a label printer (when there is already a label on each product)

Of course #1013 could also use the the scanned specific barcode as a kind of "alias" for the stock entry label barcode then, but what you said is just not correct: A barcode with weight including only identifies a stock entry uniquely unless two or more items with the very same weight are bought.

In general, tracking/using more information is always more hassle than not doing it - in any case, stock entry labels solves this problem pretty easily right now.

&nbsp;
> but I guess such a userfield won't be used with prices and energy value

Of course not, Grocy can't magically guess anything based on Userfields. Those are just fields intended for adding custom information. Like attaching the weight to each Pack / stock entry of Borsch, nothing more or less.

---
So let's conclude this clutter:
- Wildcard / weight including barcodes are a duplicate of #1013
- Being able to define the energy value in another unit than QU stock is a duplicate of #2252
- Entering the single price on [/purchase](https://demo-prerelease.grocy.info/purchase) in a different unit than selected is misleading, so won't be added/changed
- Stock entry specific conversion factors also won't be added, mainly because another complexity layer of the already complex unit conversion handling system is not what I want for the target use case of Grocy
  - Using "rough" unit conversions or practically sufficient for varying-weighted products like fruits, in relation to having a sensible-correct average price and so on
  - In conclusion just another duplicate of #2171
- Stock entry labels can be used to uniquely identify a stock entry right now
  - #1013 could extend this to use the specific scanned barcode as an alias for the corresponding stock entry, with the limitation that this will break when having > 1 stock entry with the exact same weight for this example

**Comment by berrnd** [comment]
> and their ERP supports it, I don't see why the users of Grocy can't do just the same thing.

Two different systems with whole different target use cases I guess, otherwise they would use Grocy or you should use theirs. If you're still new to and just explore Grocy at that's a breaking point for you, I personally would just skip to the next one on your list.

&nbsp;
> while borsch has a sticker per portion

Then simply scan/enter/track each portion on it's own and only the varying price / rough fix (global or product specific) conversion factor is the other open point, just like for the bananas example.

Let's make a real number example when the weight variation is even about 25% as in your last example. Say you buy 307, 364 and 377 g Packs and the rough fixed conversion is based on this, so one Pack = 349 g (`(307 + 364 + 377) / 3`) and for the sake of simplicity one Gram = 0.01 MONEY:

Average price per Gram when using the exact weight rounded to 4 decimals (without surprise): 0.0100 MONEY

Average price per Gram when using the rough conversion factor rounded to 4 decimals: 0.0100 MONEY
(`((3.07 / 349) + (3.64 / 349) + (3.77 / 349)) / 3`)

So based on that, let's take once again a statement from you:
_It's not an option, because it will provide wrong results_
=> Sure, true, because `0.01 != 0.01000955109`.

And one from me:
_Using "rough" unit conversions or practically sufficient for varying-weighted products like fruits_
=> Also true, because `0.01 == 0.01`.

**Comment by berrnd** [comment]
> without trying to understand.

If claiming that, you should try better beforehand. All that here doesn't look to me that you've tried to understand what I pointed out above, so I agree: ridiculous. Grocy pretty obviously doesn't seem to fit for you, why even bother any longer with it, I don't get it.

&nbsp;
> and that's what will be shown on the plot

It's not, when using the mentioned "rough average" conversion factor + a suitable "Quantity unit for prices" product option + sensible rounding / decimal places. But I've already given up and not going to waste any more time on that here, everything relevant has been said on this topic.
